### PR TITLE
Add Linking the .NET Libraries doc

### PR DESCRIPTION
**Pull request by eerhardt**

### CODE REVIEW

File: accepted/2020/linking-libraries.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+There are a number of work items the .NET Libraries Team needs to undertake in .NET 5. In general they fall into the two buckets outlined above:
+
+1. Linker-safety
+2. Smaller Size
+
+### Linker-safety
+
+Ensuring required code is preserved by the linker is not easy. .NET has a very rich Reflection capability that enables dynamic invocation of code. There are many cases where a static analysis tool (like the ILLinker) is not able to know which code is required and which isn't. For example, a .NET application could read a Type name from a file and create a new instance of that Type through Reflection. When the linker analyzes the application statically, it doesn't see any hard references to that Type and removes it. But when the application is executed, an error occurs becuase the Type is not in the application.
+
+To help the developer know which parts of an application or library the linker is not able to analysis correctly the linker is adding features to warn developers when they are calling linker-unfriendly methods. See the [Linker Reflection Flow Design](https://github.com/mono/linker/blob/master/docs/design/reflection-flow.md) for more information.
```
Comment (vitek-karas) [inline accepted/2020/linking-libraries.md]:
```suggestion
To help the developer know which parts of an application or library the linker is not able to analyze correctly the linker is adding features to warn developers when they are calling linker-unfriendly methods. See the [Linker Reflection Flow Design](https://github.com/mono/linker/blob/master/docs/design/reflection-flow.md) for more information.
```

File: accepted/2020/linking-libraries.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+#### Serialization
+
+An especially hard situation for the linker to understand is code that uses Reflection to (de)serialize object graphs to and from some external representation. Examples in the .NET Libraries include (but are not limited to):
+
+- `XmlSerializer`
+- `JsonSerializer`
+- `EventSource.Write<T>()`
+
+One reason this is hard is that every serialization mechanism has its own rules about what it looks for:
```
Comment (vitek-karas) [inline accepted/2020/linking-libraries.md]: I would specifically mention "serialization and deserialization" - for example I often think about "serialization" as the way to write objects to text. I don't make the connection that most people mean also "deserialization" which is the opposite. And the problems are specific to serialize/deserialize (.ctor problem is deserialize only for example).

File: accepted/2020/linking-libraries.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    - Another possibility would be a switch to make exception/resource messages smaller in the libraries. See https://github.com/dotnet/runtime/issues/34057 for more information.
+- EventSource
+    - On some devices the EventSource logic is unnecessary because it has no place to write the events. We could introduce a feature switch to remove EventSource code/calls.
+
+#### Removing Embedded Descriptor Files
+
+There are places in the .NET Libraries where we use private Reflection to call between assemblies. One reason for this is layering. If Assembly A references Assembly B, and Assembly B needs to call a method in Assembly A, it can't reference it directly. We've solved that with a Reflection call.
+
+In order to ensure the linker doesn't remove the private method, we've added "Embedded Descriptor Files" (or ILLinkTrim.xml in the `dotnet/runtime` repo) which get embedded into the built assembly. These tell the linker not to trim the specified method/type. The problem is that the linker never trims the specified method/type, even if the callsite of the Reflection is not needed.
```
Comment (vitek-karas) [inline accepted/2020/linking-libraries.md]:
```suggestion
In order to ensure the linker doesn't remove the private method, we've added "Embedded Descriptor Files" (or `ILLinkTrim.xml` in the `dotnet/runtime` repo) which get embedded into the built assembly. These tell the linker not to trim the specified method/type. The problem is that the linker never trims the specified method/type, even if the callsite of the Reflection is not needed.
```

**Review by vitek-karas — COMMENTED** [review]
Really good writeup !!! (really good!)

I don't see anything important missing. 

Maybe we could explicitly mention that some parts of the framework might be really problematic and might never work with linker well (unlike serializers which are problematic, but we want to make them work) - example could be DI (without redesign of the APIs), `Expression.Linq`, .... But I'm not sure this has place in the doc - it's more about setting expectations... so maybe we should not put it here. On the other hand in theory we can make anything work with enough changes.

File: accepted/2020/linking-libraries.md (hunk 2 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md]: fixed.

**Review by vitek-karas — APPROVED** [review]

File: accepted/2020/linking-libraries.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+- `EventSource.Write<T>()`
+
+One reason this is hard is that every serialization and deserialization mechanism has its own rules about what it looks for:
+- What are acceptable constructors to create the object?
+- What properties (or fields) are inspected?
+- Does handle derived types?
+
+As of this writing, there is not a bullet proof solution for the linker to work in all serialization cases. See https://github.com/mono/linker/issues/1087 for a proposal on what it would take to make `JsonSerializer` linker friendly.
+
+One solution that would solve most of the issues is to create [Source Generators](https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/) for each supported serializer. This would remove the use of Reflection, and would allow the linker to more easily see which code must be preserved.
```
Comment (jkotas) [inline accepted/2020/linking-libraries.md]: Worth calling that this is not in schedule for .NET 5, and thus these serializers won't work well with the IL linker in .NET 5?

File: accepted/2020/linking-libraries.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+ILLink is capable of saving the dependency graph when it trims an application. There is a tool, [Linker Analyzer](https://github.com/mono/linker/tree/d3eb4ea8179516b0554a3d3089c71062496bd6f0/src/analyzer), that can load this graph and allows a developer to analyze it. Example questions that can be answered are:
+
+1. What are the biggest types/methods left after trimming?
+2. What is the call chain that is keeping type/method `foo` from being trimmed?
+
+Today the Linker Analyzer is a console application. One enhancement we can make is to put the core logic into a NuGet package, and allow a developer to analyze their application using the [.NET Interactive](https://github.com/dotnet/interactive/blob/master/README.md) notebook experience. Advantages of this experience are:
```
Comment (jkotas) [inline accepted/2020/linking-libraries.md:142]: The best tool that I have seen for this was https://github.com/microsoft/perfview . The observation is that profiling size is very similar to profiling cycles. We had capability to load the method size data into perfview in .NET Native days.

> biggest types/methods left after trimming?

This is equivalent of biggest inclusive/exclusive time

> What is the call chain that is keeping type/method `foo` from being trimmed?

This is call graph.

File: accepted/2020/linking-libraries.md [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+In the context of trimming applications, there are two main concerns:
+
+1. Any code that is used by the application must be preserved in the application.
+2. Any code that is not used by the application can be removed from the application.
+
+Note the difference in "must" and "can" between those two concerns. An application that works is preferred over a smaller application that doesn't. Therefore it is critical that any necessary code must be preserved in the application.
+
+## Scenarios and User Experience
+
+Developers can run the ILLinker on their .NET application and be confident that the resulting application works correctly, while meeting size goals.
```
Comment (marek-safar) [inline accepted/2020/linking-libraries.md]: I'd suggest rewording this a bit, to me this now sounds like the ILLinker will be additional work for developers.

Also `while meeting size goals.` might be too strong wording.

File: accepted/2020/linking-libraries.md [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+This is especially important for applications that need to be downloaded over the internet, for example Xamarin and Blazor applications. End users on a slow or limited network will be less likely to install and use applications that are too large.
+
+The [ILLinker](https://github.com/mono/linker) is a tool that is able to take an application and trim the classes and methods that are not used by the application. In many scenarios, this results in a significantly smaller packaged application.
+
+In .NET 5 we will modify the [.NET Libraries](https://github.com/dotnet/runtime/tree/master/src/libraries) to work well with the ILLinker. With .NET 5, our goal is to use the .NET Libraries for all .NET applications. In previous releases, Xamarin and Blazor applications have been using the Mono libraries, which have been written with size in mind. However, the .NET Libraries have traditionally been concerned with throughput and not size.
```
Comment (marek-safar) [inline accepted/2020/linking-libraries.md]: I'd prefer this to be changed to describe more the long term strategy than promising that .NET5 will be size optimal release which we know that it won't be.

File: accepted/2020/linking-libraries.md [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+In the context of trimming applications, there are two main concerns:
+
+1. Any code that is used by the application must be preserved in the application.
```
Comment (marek-safar) [inline accepted/2020/linking-libraries.md]:
```suggestion
1. Any code and data that are used by the application must be preserved in the self-contained application.
```

File: accepted/2020/linking-libraries.md [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+2. Any code that is not used by the application can be removed from the application.
```
Comment (marek-safar) [inline accepted/2020/linking-libraries.md]:
```suggestion
2. Any code or dependencies that are not used by the application can be removed from the self-contained application.
```

File: accepted/2020/linking-libraries.md [hunk 10]
```diff
@@ -0,0 +1,152 @@
+# Linking the .NET Libraries
+
+When publishing a self-contained application, the .NET Core runtime is bundled together with the application. This bundling adds a significant amount of content to the packaged application. Depending on the complexity of the application, only a subset of the runtime is required to run the application. These unused parts of the runtime are unnecessary and can be trimmed from the packaged application.
```
Comment (marek-safar) [inline accepted/2020/linking-libraries.md:3]: Please also explain the impact/intention for the application using the shared framework

File: accepted/2020/linking-libraries.md [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+1. Ensure all required code is preserved by the linker.
+2. Allow Xamarin and Blazor applications to use the .NET Libraries and be as small as possible.
+
+### Non-Goals
+
+1. Optimizing application size in other form factors.
+
+## Design
+
+There are a number of work items the .NET Libraries Team needs to undertake in .NET 5. In general they fall into the two buckets outlined above:
```
Comment (marek-safar) [inline accepted/2020/linking-libraries.md]: I'd suggest removing .NET 5 references from this doc

File: accepted/2020/linking-libraries.md [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+
+There are features in the .NET Libraries that will be very expensive, and in some cases virtually impossible, to make 100% linker-safe. A few examples are:
+- `System.ComponentModel.Composition`, `System.Composition`, and other dependency injection style libraries
+    - Contains features such as loading "parts" from a directory
+- `Microsoft.CSharp` and `System.Linq.Expressions`
+    - Use Reflection extensively to implement `dynamic` and expression trees
+- `System.Resources.ResourceManager`
+    - Resource files can have ResourceReader Type names embedded in them, which `ResourceManager` creates dynamically.
+
```
Comment (marek-safar) [inline accepted/2020/linking-libraries.md:57]: It might be worth to call explicitly System.Reflection.Emit too

File: accepted/2020/linking-libraries.md [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+For these expensive cases, we will need to evaluate if the amount of effort required to make the feature linker-safe will add enough value to the scenarios we are targetting. For example, it is unlikely many developers are using `System.Composition` in Xamarin applications.
+
+To help the developer know which parts of an application or library the linker is not able to analyze correctly the linker is adding features to warn developers when they are calling linker-unfriendly methods. See the [Linker Reflection Flow Design](https://github.com/mono/linker/blob/master/docs/design/reflection-flow.md) for more information.
+
+#### Adding linker annotations
```
Comment (marek-safar) [inline accepted/2020/linking-libraries.md]:
```suggestion
#### Adding dynamic access annotations
```

File: accepted/2020/linking-libraries.md [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+
+2. In some cases the usage is actually safe. For example, if an "Equals" method is written by retrieving all the fields on a type and looping over them comparing each field. If a field is unused, and therefore trimmed by the linker, there is no need to compare that field for equality. In these cases we can suppress the linker warning using the `[UnconditionalSuppressMessage]` attribute.
+
+3. In some cases the API itself is not able to be linker friendly. In these cases we don't want to suppress the warning completely, but instead pass the warning to the caller of the API. In these cases we can use the `[RequiresUnreferencedCode]` attribute. This attribute tells the linker to not warn about this method, but instead warn about any callers of this method.
+
+A high-level goal is that the .NET Libraries are clean of linker warnings. This is one level of confidence we can have that .NET applications will be linker safe.
+
+#### Serialization
```
Comment (marek-safar) [inline accepted/2020/linking-libraries.md:77]: Why to explicitly call out serialization only? What about other large framework features should they be described in the same level of detail?

File: accepted/2020/linking-libraries.md [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+
+As of this writing, there is not a bullet proof solution for the linker to work in all serialization cases. See https://github.com/mono/linker/issues/1087 for a proposal on what it would take to make `JsonSerializer` linker friendly.
+
+One solution that would solve most of the issues is to create [Source Generators](https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/) for each supported serializer. This would remove the use of Reflection, and would allow the linker to more easily see which code must be preserved.
+
+### Smaller Size
+
+To make applications smaller, we have to let the linker know what code is not necessary and can be removed. A lot of times it is obvious that if a method isn't called, it can be trimmed. But there are places where unnecessary code looks like it is being called.
+
+#### Adding appropriate feature switches
```
Comment (marek-safar) [inline accepted/2020/linking-libraries.md]:
```suggestion
#### Removing Framework Features
```

File: accepted/2020/linking-libraries.md [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+
+## Testing
+
+Testing the linked applications falls into two buckets:
+1. Does the application still work?
+2. Is the application smaller?
+
+In general, using unit tests to ensure a linked application still works isn't a great testing strategy. Unit tests, by definition, try exercising all the code in the library. If all the code in the library is being exercised, the linker won't be able to trim any of it. Thus if nothing is trimmed, you aren't confirming that the library still works when not all of the code is present.
+
+Instead of using unit tests, we should create dedicated applications to test that the libraries are linker safe. One set of applications will be higher-level applications that we want to measure the size over time. Another set will target specific APIs that needed to be annotated to ensure the annotations and the linker are working correctly.
```
Comment (marek-safar) [inline accepted/2020/linking-libraries.md:160]: Could this be rewritten to be more actionable? 
	- How should target specific APIs test look like?
	- Where they will be located?
	- How they will be run?
	- Who will be responsible for writing the test(s)?

File: accepted/2020/linking-libraries.md [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+
+As we are measuring size over time, we would not like to get "noise" in the sizes to cause false positives. For example, if an application size gets larger by ~10 KB, we don't necessarily want the test to fail. And conversely, if an application size suddenly get 50% larger (or smaller), we probably do want to alert someone - as an anomaly was detected and it is possibly a bug somewhere.
```
Comment (marek-safar) [inline accepted/2020/linking-libraries.md:164]: @billwert is working on this and has design/proposal for this kind of testing. It might be better to link it from here instead.

File: accepted/2020/linking-libraries.md (hunk 5 above)
Comment (marek-safar) [inline accepted/2020/linking-libraries.md]: I'm not sure we should cover this in detail here as we don't have a good plan right. We need much better tooling in general not only for us but for any developer. I'd suggest describing the intentions here in general and say the size tooling spec will be follow up.

File: accepted/2020/linking-libraries.md (hunk 14 above)
Comment (jkotas) [reply accepted/2020/linking-libraries.md:77]: https://github.com/dotnet/runtime/tree/master/src/libraries/Microsoft.Extensions.DependencyInjection is another big one

File: accepted/2020/linking-libraries.md [hunk 18]
```diff
@@ -0,0 +1,152 @@
+# Linking the .NET Libraries
+
+When publishing a self-contained application, the .NET Core runtime is bundled together with the application. This bundling adds a significant amount of content to the packaged application. Depending on the complexity of the application, only a subset of the runtime is required to run the application. These unused parts of the runtime are unnecessary and can be trimmed from the packaged application.
+
+This is especially important for applications that need to be downloaded over the internet, for example Xamarin and Blazor applications. End users on a slow or limited network will be less likely to install and use applications that are too large.
+
+The [ILLinker](https://github.com/mono/linker) is a tool that is able to take an application and trim the classes and methods that are not used by the application. In many scenarios, this results in a significantly smaller packaged application.
```
Comment (stephentoub) [inline accepted/2020/linking-libraries.md:9]: Complete aside, but are there any plans to move https://github.com/mono/linker to be in the dotnet org as https://github.com/dotnet/linker?

File: accepted/2020/linking-libraries.md (hunk 12 above)
Comment (stephentoub) [reply accepted/2020/linking-libraries.md:57]: I'm curious how emit plays into this?  I'd have expected it to not really impact linking, since by definition it's generating code at run time to be used at run time, and the linker doesn't come into play?

File: accepted/2020/linking-libraries.md [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+- `System.ComponentModel.Composition`, `System.Composition`, and other dependency injection style libraries
+    - Contains features such as loading "parts" from a directory
+- `Microsoft.CSharp` and `System.Linq.Expressions`
+    - Use Reflection extensively to implement `dynamic` and expression trees
+- `System.Resources.ResourceManager`
+    - Resource files can have ResourceReader Type names embedded in them, which `ResourceManager` creates dynamically.
+
+For these expensive cases, we will need to evaluate if the amount of effort required to make the feature linker-safe will add enough value to the scenarios we are targetting. For example, it is unlikely many developers are using `System.Composition` in Xamarin applications.
+
+To help the developer know which parts of an application or library the linker is not able to analyze correctly the linker is adding features to warn developers when they are calling linker-unfriendly methods. See the [Linker Reflection Flow Design](https://github.com/mono/linker/blob/master/docs/design/reflection-flow.md) for more information.
```
Comment (stephentoub) [inline accepted/2020/linking-libraries.md:60]: Should we also call out a case like `System.Private.DataContractSerialization`, where there's so much internal reflection that we simply disable linking of it?
```XML
    <!-- Too much private reflection. Do not bother with trimming -->
    <ILLinkTrimAssembly>false</ILLinkTrimAssembly>
```
i.e. if you use any of the library, you get the whole thing.

File: accepted/2020/linking-libraries.md [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+
+#### Adding linker annotations
+
+With the addition of the Linker Reflection Flow features above, we will need to add attributes to the .NET Library methods that rely on Reflection. At a high-level, we can run the linker on the libraries and look at the warnings it produces. We then need to analyze the correct fix for the linker warning. Possibilities are:
+
+1. Add a `DynamicallyAccessedMembers` attribute to the code, to tell the linker which members need to be accessed dynamically.
+    - This attribute is similar in nature to the nullable annotations that were added in C# 8 in that they are both viral. `DynamicallyAccessedMembers` needs to be applied from the actual usage of Reflection all the way through the call tree to the public API that a developer calls passing in a Type.
```
Comment (stephentoub) [inline accepted/2020/linking-libraries.md]: > all the way through the call tree to the public API that a developer calls passing in a Type

Or more specifically to the point where the type information is provided, right? i.e. we might use one of these in a private method B that's called from another private method A, and A could declare everything safe such the virality ends at that point in that specific call chain?

File: accepted/2020/linking-libraries.md [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+
+2. In some cases the usage is actually safe. For example, if an "Equals" method is written by retrieving all the fields on a type and looping over them comparing each field. If a field is unused, and therefore trimmed by the linker, there is no need to compare that field for equality. In these cases we can suppress the linker warning using the `[UnconditionalSuppressMessage]` attribute.
+
+3. In some cases the API itself is not able to be linker friendly. In these cases we don't want to suppress the warning completely, but instead pass the warning to the caller of the API. In these cases we can use the `[RequiresUnreferencedCode]` attribute. This attribute tells the linker to not warn about this method, but instead warn about any callers of this method.
+
```
Comment (stephentoub) [inline accepted/2020/linking-libraries.md:74]: Do you want to mention DynamicDependency?

File: accepted/2020/linking-libraries.md [hunk 22]
```diff
@@ -0,0 +1,152 @@
...
+### Smaller Size
+
+To make applications smaller, we have to let the linker know what code is not necessary and can be removed. A lot of times it is obvious that if a method isn't called, it can be trimmed. But there are places where unnecessary code looks like it is being called.
+
+#### Adding appropriate feature switches
+
+One mechanism we are adding that will allow us to conditionally remove code from applications is [feature switches](./feature-switch.md). Using this mechanism, an SDK (like Xamarin) or the developer can decide if a feature that would normally be available in the libraries can be removed, both at runtime and during linking. An example of an existing feature switch we have today (that was introduced for other reasons than application size) is [InvariantGlobalization](https://docs.microsoft.com/en-us/dotnet/core/run-time-config/globalization). We can add more switches to the libraries to allow large, optional pieces of code to be removed by the linker.
```
Comment (stephentoub) [inline accepted/2020/linking-libraries.md]: > that was introduced for other reasons than application size

Arguably at the end of the day it was still about application size, in that we could have distributed ICU or made it a hard-and-fast requirement of the runtime, but both of those would have increased size on OSes like Alpine that otherwise didn't have it.

But that's nit-picking; what you have is fine.

File: accepted/2020/linking-libraries.md [hunk 23]
```diff
@@ -0,0 +1,152 @@
...
+
+There are 2 existing cases here worth mentioning:
+1. Types only used for Debugging
+2. Types used for COM support
+
+In both of these cases, the code is sometimes needed, and sometimes isn't. If you are never going to debug your application (for example, the application is deployed to a device that can't be debugged), it isn't necessary to keep these types in the application. A similar situation exists with COM. Since COM is not IL, the linker isn't able to see if some code in the application is going to use COM or not.
+
+To solve these, we can define a [`feature switch`](#Adding-appropriate-feature-switches) for each of these scenarios. An application (or SDK) can default these to on or off, as appropriate. For example, a Xamarin iOS application can safely turn COM support off.
```
Comment (stephentoub) [inline accepted/2020/linking-libraries.md]: Are the COM-related APIs kept around today even on Unix because of the S.P.Corelib ILLinkTrim.xml file?  If yes, will that matter for the goaled-application types anyway, since they'll be targeting the mono runtime and its Corelib?

Separately, if we are thinking about coreclr as part of this effort, have we given any thought to how we might untangle the dependencies introduced by the native runtime pieces?  We generate an ILLinkTrim.xml file that roots anything that might ever be used from native code, regardless of whether that code could actually end up being used.

File: accepted/2020/linking-libraries.md [hunk 24]
```diff
@@ -0,0 +1,152 @@
...
+4. When a new feature is added to the Analyzer, we just need to create a public API. We don't need to create command line arguments.
+5. We have the option of visuals and more interactive output - for example hyperlinks. In a console application, you are stuck with text.
+
+## Testing
+
+Testing the linked applications falls into two buckets:
+1. Does the application still work?
+2. Is the application smaller?
+
+In general, using unit tests to ensure a linked application still works isn't a great testing strategy. Unit tests, by definition, try exercising all the code in the library. If all the code in the library is being exercised, the linker won't be able to trim any of it. Thus if nothing is trimmed, you aren't confirming that the library still works when not all of the code is present.
```
Comment (stephentoub) [inline accepted/2020/linking-libraries.md]:
```suggestion
In general, using a library's unit tests to ensure a linked application still works isn't a great testing strategy. Unit tests, by definition, try exercising all the code in the library. If all the code in the library is being exercised, the linker won't be able to trim any of it. Thus if nothing is trimmed, you aren't confirming that the library still works when not all of the code is present.  Similarly, if we try to run a library's unit tests on a linked build of that library as included in an application, it's likely that a large percentage of the library's tests will fail, as they were targeting APIs that were linked away having not been required for the application.
```

**Review by stephentoub — APPROVED** [review]
Thanks for writing this down.

File: accepted/2020/linking-libraries.md (hunk 5 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md:142]: Is it possible to use perfview today for this? If we put the information in the correct format, will perfview know how to show it? Or would work need to be done to perfview?

File: accepted/2020/linking-libraries.md (hunk 6 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md]: > this now sounds like the ILLinker will be additional work for developers

It can be, right? For example if a developer is using reflection in their application, they may need to add annotations to their code.

I can reword this sentence to:

> The ILLinker can be executed on a .NET application and the developer will be confident that the trimmed application works correctly, while reducing its final size.

File: accepted/2020/linking-libraries.md (hunk 7 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md]: Good point. I'll reword any "time" considerations to not be version specific.

File: accepted/2020/linking-libraries.md (hunk 9 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md]: I'm not sure I follow the `or dependencies` part. To me dependencies are code as well. I'd also like to keep these two sentences symmetrical, to help drive the point of `must` and `can` below.

File: accepted/2020/linking-libraries.md (hunk 10 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md:3]: I'm not sure I know the intention for framework-dependent apps. Are we intending to be able to trim those?

File: accepted/2020/linking-libraries.md (hunk 11 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md]: Agreed.

File: accepted/2020/linking-libraries.md (hunk 12 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md:57]: A place it comes into play is if you are emitting a call to a method/property that is trimmed away. The linker may think that method is not needed since it isn't called statically.

File: accepted/2020/linking-libraries.md (hunk 10 above)
Comment (jkotas) [reply accepted/2020/linking-libraries.md:3]: Yes - in the fullness of time

File: accepted/2020/linking-libraries.md (hunk 14 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md:77]: My reasoning was that serialization seemed to be a common enough pattern used across many features, and in some cases it isn't immediately obvious that a feature is serializing objects - e.g. `EventSource`. It isn't a single feature, but instead a class/category of features that is problematic for the linker.

File: accepted/2020/linking-libraries.md (hunk 17 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md:164]: @billwert - I haven't seen your proposal, can you share a link that I can add?

File: accepted/2020/linking-libraries.md (hunk 18 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md:9]: @marek-safar @vitek-karas - thoughts?

File: accepted/2020/linking-libraries.md (hunk 5 above)
Comment (jkotas) [reply accepted/2020/linking-libraries.md:142]: I guess there may be bugfixes required due to bit-rot, but the code is still there. 

In .NET Native days, there was a separate tool called PDBScope that generated .xml from the .PDB. This .XML file is what perfview read and displayed.

https://github.com/microsoft/perfview/search?q=PDBScope

File: accepted/2020/linking-libraries.md (hunk 19 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md:60]: That setting is actually only used in our build. It doesn't affect end-user's applications that happen to use `System.Private.DataContractSerialization`. That might be a problem.

If we wanted the end-user's application to keep the whole assembly, we would probably need to add an xml file to it that says to keep the whole assembly. But I'd hope we can annotate the code enough to make it linker-safe.

File: accepted/2020/linking-libraries.md (hunk 20 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md]: Good point. I've reworded this:

> `DynamicallyAccessedMembers` needs to be applied from the actual usage of Reflection all the way through the call tree to the location that specifies the Type, which may be a public API.

File: accepted/2020/linking-libraries.md (hunk 5 above)
Comment (jkotas) [reply accepted/2020/linking-libraries.md:142]: You can get a feel for what it looks like by opening perfview on a directory (Menu: Size -> DirectorySize) like `c:\windows`. It will give you the profiler view for analyzing file sizes inside a directory structure.

File: accepted/2020/linking-libraries.md (hunk 22 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md]: I've removed the aside. I agree the main purpose of the feature switch was size reasons.

File: accepted/2020/linking-libraries.md (hunk 23 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md]: There are COM-only types outside of CoreLib:

https://github.com/dotnet/runtime/blob/e77572ffccc566186f47207f3c5475533c87538e/src/libraries/Microsoft.CSharp/src/ILLinkTrim.xml#L8

https://github.com/dotnet/runtime/blob/e77572ffccc566186f47207f3c5475533c87538e/src/libraries/System.Runtime.InteropServices/src/ILLinkTrim.xml#L3-L4

> have we given any thought to how we might untangle the dependencies introduced by the native runtime pieces?

My only thoughts have been the feature switch in the ILLinkTrim.xml file idea above that would allow application to remove them dynamically. Maybe @vitek-karas or @marek-safar have more thoughts here?

File: accepted/2020/linking-libraries.md (hunk 4 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md]: I've added:

> However, Source Generators are a new feature and are potentionally expensive to create. As such, it may take some time before they are used broadly.

File: accepted/2020/linking-libraries.md (hunk 23 above)
Comment (jkotas) [reply accepted/2020/linking-libraries.md]: > any thought to how we might untangle the dependencies introduced by the native runtime pieces?

Lift more of the unmanaged logic into C# as necessary, use explicitly passed in callbacks instead of calling functions by name from VM.

File: accepted/2020/linking-libraries.md (hunk 5 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md]: I've listed the high level intentions in this section, and moved the more detailed information into a `Potential Size Analysis Tooling` sub-section.

Comment (eerhardt) [reply accepted/2020/linking-libraries.md:142]: Thanks, I've added PerfView to the list of potential tooling available.

File: accepted/2020/linking-libraries.md (hunk 16 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md:160]: > Who will be responsible for writing the test(s)?

The .NET Library developers will write the target specific APIs test, as it is API specific.

As for the higher-level applications, I think that is up for debate whether it is the perf team, the application-model team (Xamarin, Blazor, etc), the Libraries team, or someone else.

> Where they will be located?

I think a decent place for the higher-level applications would be https://github.com/dotnet/performance. It has the ability to do the "measure over time" capabilities that I mention in the doc. It would also keep application-model specific dependencies out of dotnet/runtime.

As for the targeted API tests, I was assuming they would be in dotnet/runtime, since they are tied to the APIs.

File: accepted/2020/linking-libraries.md (hunk 19 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md:60]: I've logged https://github.com/dotnet/runtime/issues/36154 for this issue.

File: accepted/2020/linking-libraries.md (hunk 12 above)
Comment (stephentoub) [reply accepted/2020/linking-libraries.md:57]: > emitting a call to a method/property that is trimmed away

Sure, but to use the emit APIs to do so, you need to use reflection to get the relevant members so that you can pass in the MemberInfo to invoke, so it would seem to be the same reflection case.

Maybe I'm splitting hairs. Doesn't really matter.

Comment (marek-safar) [reply accepted/2020/linking-libraries.md:57]: I think you are underestimating the SRE scope. For example, you can use SRE with tokens which you can get from handles, signatures or even stream of bytes. You can also generate new assemblies (we don't have a good concept for that in linker).

File: accepted/2020/linking-libraries.md (hunk 14 above)
Comment (marek-safar) [reply accepted/2020/linking-libraries.md:77]: If you meant to include serialization only as example then I think that's fine but it should be more obvious from the text

File: accepted/2020/linking-libraries.md (hunk 16 above)
Comment (marek-safar) [reply accepted/2020/linking-libraries.md:160]: > As for the targeted API tests

Where and how they will be written when they cannot be xunit based or use any testing framework?

I'd focus on unit test only in this doc but define them well (see the questions I asked they were all about unit tests)

File: accepted/2020/linking-libraries.md (hunk 18 above)
Comment (marek-safar) [reply accepted/2020/linking-libraries.md:9]: No, I don't see any benefit of that work. Why?

File: accepted/2020/linking-libraries.md (hunk 23 above)
Comment (marek-safar) [reply accepted/2020/linking-libraries.md]: > have we given any thought to how we might untangle the dependencies introduced by the native runtime pieces?

It depends. The XML descriptors are actually a good fit for this. If we know the runtime code will be 99% time present we just add it to XML (e.g. string constructor) for the rest it's a combination of using `DynamicDependencyAttribute` (aka PreserveDependency) and for large chunks of native code moving them to manage for better likability.

File: accepted/2020/linking-libraries.md (hunk 9 above)
Comment (marek-safar) [reply accepted/2020/linking-libraries.md]: The dependencies in this context are for example "facade" assemblies, unused Android resources or unmanaged native libraries.

File: accepted/2020/linking-libraries.md (hunk 23 above)
Comment (stephentoub) [reply accepted/2020/linking-libraries.md]: > The XML descriptors are actually a good fit for this

I'm surprised by that. From what I've seen, the XML isn't a great fit because a non-trivial amount of what's automatically kept isn't actually used by any given app.  I could be wrong, of course.

File: accepted/2020/linking-libraries.md (hunk 12 above)
Comment (stephentoub) [reply accepted/2020/linking-libraries.md:57]: Saving assemblies isn't supported in core. And the vast majority use case doesn't use those niche features you describe. I would be surprised if anyone using the linker with .NET 5 breaks because of them, but sure, it's possible. I would not be surprised if someone breaks because they emit a call via a MethodInfo to something that gets removed, but that's about the MethodInfo rather than the emit.

Comment (stephentoub) [reply accepted/2020/linking-libraries.md:57]: (For the purposes of this doc, of course, it doesn't matter and is fine to call out if you think it's important.  I'm just questioning it from a prioritization perspective.)

File: accepted/2020/linking-libraries.md (hunk 23 above)
Comment (marek-safar) [reply accepted/2020/linking-libraries.md]: > by the native runtime pieces

I was specifically referring to core runtime dependencies. There is a certain amount of core types/methods which runtime always requires (e.g. System.Object, System.Exception, etc) to even start. I don't think we need to make that conditional and listing such core dependencies in S.P.C XML descriptor is fine.

Comment (stephentoub) [reply accepted/2020/linking-libraries.md]: Sure, there's a subset that's always used.

File: accepted/2020/linking-libraries.md (hunk 12 above)
Comment (jkotas) [reply accepted/2020/linking-libraries.md:57]: >  you can use SRE with tokens which you can get from handles, signatures or even stream of bytes.

These sound like special instances of dynamic loading (of assemblies that the linker did not know about). It is same category as `Assembly.Load(byte[])` or `Assembly.LoadFile`. Dynamic loading is not compatible with linker by design. Maybe mention dynamic loading as an example of feature that won't work with linker, ever?

File: accepted/2020/linking-libraries.md (hunk 18 above)
Comment (jkotas) [reply accepted/2020/linking-libraries.md:9]: It would be for the same reason why we have moved e.g. ASP.NET from the aspnet org to dotnet org. The primary reasons are consistency and public perception.

Other more technical reasons include permission management or complications moving issues between repos when they are in different orgs.

File: accepted/2020/linking-libraries.md (hunk 12 above)
Comment (eerhardt) [reply accepted/2020/linking-libraries.md:57]: > Maybe mention dynamic loading as an example of feature that won't work with linker, ever?

Added.
